stl/inc/xutility

```diff
@@ -866,8 +866,33 @@ namespace ranges {
         template <class _Ty>
         concept _Can_deref = requires(_Ty&& __t) { *static_cast<_Ty&&>(__t); };
 
+        struct _Must_have_ADL_found_iter_move_or_be_dereferenceable {};
+
         class _Cpo {
         private:
+            template <class _Ty>
+            _NODISCARD static auto _Choose_return_type() {
+                if constexpr (_Has_ADL<_Ty>) {
+                    return type_identity<decltype(iter_move(_STD declval<_Ty>()))>{}; // intentional ADL
+                } else if constexpr (_Can_deref<_Ty>) {
+                    using _Ref = decltype(*_STD declval<_Ty>());
+                    if constexpr (is_function_v<remove_reference_t<_Ref>>) {
+                        // TRANSITION, VSO-1008447, VSO-2066340
+                        // Fuse this into the type_identity<_Ref>{} branch once MSVC is fixed.

```

Comment:
It is somewhat unclear to me what this fusion will actually look like. We should follow the usual "workaround / no workaround" convention, which will both exercise the good path for non-MSVC, and make it clear how to eliminate the bad path in the future.

Reply:
Ah, thank you! When the return type is a function rvalue reference, the function call expression should still be an lvalue, although MSVC is still buggy on this. I was confused before, and now I think this branch should be removed once MSVC has the bug fixed.